src/transformers/models/jukebox/modeling_jukebox.py

```diff
@@ -0,0 +1,2694 @@
+            _random_codebook = codes[torch.randperm(codes.shape[0])][:codebook_dim]
+
+            # Update centres
+            old_codebook = self.codebook
+            self.codebook_sum = mu * self.codebook_sum + (1.0 - mu) * _codebook_sum  # w, codebook_dim
+            self.codebook_elem = mu * self.codebook_elem + (1.0 - mu) * _codebook_elem  # codebook_dim
+            usage = (self.codebook_elem.view(codebook_dim, 1) >= self.threshold).float()
+            self.codebook = (
+                usage
+                * (self.codebook_sum.view(codebook_dim, codebook_width) / self.codebook_elem.view(codebook_dim, 1))
+                + (1 - usage) * _random_codebook
+            )
+            _codebook_prob = _codebook_elem / torch.sum(
+                _codebook_elem
+            )  # latent_states_onehot.mean(dim=-1)  # prob of each bin
+            entropy = -torch.sum(_codebook_prob * torch.log(_codebook_prob + 1e-8))  # entropy ie how diverse
+            used_curr = (_codebook_elem >= self.threshold).sum()
+            usage = torch.sum(usage)
+            dk = torch.norm(self.codebook - old_codebook) / np.sqrt(np.prod(old_codebook.shape))
+        return dict(entropy=entropy, used_curr=used_curr, usage=usage, dk=dk)
+
+    def preprocess(self, hidden_states):
+        # NCT -> NTC -> [NT, C]
+        hidden_states = hidden_states.permute(0, 2, 1).contiguous()
+        hidden_states = hidden_states.view(
+            -1, hidden_states.shape[-1]
+        )  # x_en = (batch_size *L, w), k_j = (w, codebook_dim)
+
+        if hidden_states.shape[-1] == self.codebook_width:
+            prenorm = torch.norm(hidden_states - torch.mean(hidden_states)) / np.sqrt(np.prod(hidden_states.shape))
+        elif hidden_states.shape[-1] == 2 * self.codebook_width:
+            x1, x2 = hidden_states[..., : self.codebook_width], hidden_states[..., self.codebook_width :]
+            prenorm = (torch.norm(x1 - torch.mean(x1)) / np.sqrt(np.prod(x1.shape))) + (
+                torch.norm(x2 - torch.mean(x2)) / np.sqrt(np.prod(x2.shape))
+            )
+
+            # Normalise
+            hidden_states = x1 + x2
+
+        return hidden_states, prenorm
+
+    def postprocess(self, latent_states, dequantised_states, x_shape):
+        # [NT, C] -> NTC -> NCT
+        batch_size, time = x_shape
+        dequantised_states = dequantised_states.view(batch_size, time, -1).permute(0, 2, 1).contiguous()
+        latent_states = latent_states.view(batch_size, time)
+        return latent_states, dequantised_states
+
+    def quantise(self, latent_states):
+        # Calculate latent code latent_states
+        codebook_weights = self.codebook.t()
+        distance = (
+            torch.sum(latent_states**2, dim=-1, keepdim=True)
+            - 2 * torch.matmul(latent_states, codebook_weights)
+            + torch.sum(codebook_weights**2, dim=0, keepdim=True)
+        )  # (batch_size * latent_states , codebook_weights) # better help from this comment
```

Comment:
```suggestion
        )  # (batch_size * latent_states , codebook_weights)
```
Might be better on top.